// graphql-transformer-common/lib/ResolverResourceIDs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResolverResourceIDs = void 0;
class ResolverResourceIDs {
    static DynamoDBCreateResolverResourceID(typeName) {
        return `Create${typeName}Resolver`;
    }
    static DynamoDBUpdateResolverResourceID(typeName) {
        return `Update${typeName}Resolver`;
    }
    static DynamoDBDeleteResolverResourceID(typeName) {
        return `Delete${typeName}Resolver`;
    }
    static DynamoDBGetResolverResourceID(typeName) {
        return `Get${typeName}Resolver`;
    }
    static DynamoDBListResolverResourceID(typeName) {
        return `List${typeName}Resolver`;
    }
    static ElasticsearchSearchResolverResourceID(typeName) {
        return `Search${typeName}Resolver`;
    }
    static SyncResolverResourceID(typeName) {
        return `Sync${typeName}Resolver`;
    }
    static ResolverResourceID(typeName, fieldName) {
        return `${typeName}${fieldName}Resolver`;
    }
}
exports.ResolverResourceIDs = ResolverResourceIDs;
