// graphql-transformer-core/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConflictHandlerType = exports.TRANSFORM_CURRENT_VERSION = exports.TRANSFORM_BASE_VERSION = exports.TRANSFORM_CONFIG_FILE_NAME = exports.revertAPIMigration = exports.writeTransformerConfiguration = exports.readTransformerConfiguration = exports.readProjectConfiguration = exports.readProjectSchema = exports.uploadAPIProject = exports.migrateAPIProject = exports.buildAPIProject = exports.stripDirectives = exports.collectDirectivesByTypeNames = exports.collectDirectiveNames = exports.Transformer = exports.TransformerContext = exports.GraphQLTransform = exports.getAppSyncServiceExtraDirectives = void 0;
require("./polyfills/Object.assign");
const graphql_1 = require("graphql");
const TransformerContext_1 = require("./TransformerContext");
Object.defineProperty(exports, "TransformerContext", { enumerable: true, get: function () { return TransformerContext_1.TransformerContext; } });
const Transformer_1 = require("./Transformer");
Object.defineProperty(exports, "Transformer", { enumerable: true, get: function () { return Transformer_1.Transformer; } });
const GraphQLTransform_1 = require("./GraphQLTransform");
Object.defineProperty(exports, "GraphQLTransform", { enumerable: true, get: function () { return GraphQLTransform_1.GraphQLTransform; } });
const collectDirectives_1 = require("./collectDirectives");
Object.defineProperty(exports, "collectDirectiveNames", { enumerable: true, get: function () { return collectDirectives_1.collectDirectiveNames; } });
Object.defineProperty(exports, "collectDirectivesByTypeNames", { enumerable: true, get: function () { return collectDirectives_1.collectDirectivesByTypeNames; } });
const stripDirectives_1 = require("./stripDirectives");
Object.defineProperty(exports, "stripDirectives", { enumerable: true, get: function () { return stripDirectives_1.stripDirectives; } });
const amplifyUtils_1 = require("./util/amplifyUtils");
Object.defineProperty(exports, "buildAPIProject", { enumerable: true, get: function () { return amplifyUtils_1.buildProject; } });
Object.defineProperty(exports, "uploadAPIProject", { enumerable: true, get: function () { return amplifyUtils_1.uploadDeployment; } });
Object.defineProperty(exports, "migrateAPIProject", { enumerable: true, get: function () { return amplifyUtils_1.migrateAPIProject; } });
Object.defineProperty(exports, "revertAPIMigration", { enumerable: true, get: function () { return amplifyUtils_1.revertAPIMigration; } });
const transformConfig_1 = require("./util/transformConfig");
Object.defineProperty(exports, "readProjectSchema", { enumerable: true, get: function () { return transformConfig_1.readSchema; } });
Object.defineProperty(exports, "readProjectConfiguration", { enumerable: true, get: function () { return transformConfig_1.loadProject; } });
Object.defineProperty(exports, "readTransformerConfiguration", { enumerable: true, get: function () { return transformConfig_1.loadConfig; } });
Object.defineProperty(exports, "writeTransformerConfiguration", { enumerable: true, get: function () { return transformConfig_1.writeConfig; } });
Object.defineProperty(exports, "TRANSFORM_CONFIG_FILE_NAME", { enumerable: true, get: function () { return transformConfig_1.TRANSFORM_CONFIG_FILE_NAME; } });
Object.defineProperty(exports, "TRANSFORM_BASE_VERSION", { enumerable: true, get: function () { return transformConfig_1.TRANSFORM_BASE_VERSION; } });
Object.defineProperty(exports, "TRANSFORM_CURRENT_VERSION", { enumerable: true, get: function () { return transformConfig_1.TRANSFORM_CURRENT_VERSION; } });
const validation_1 = require("./validation");
__exportStar(require("./errors"), exports);
__exportStar(require("./util"), exports);
function getAppSyncServiceExtraDirectives() {
    return graphql_1.print(validation_1.EXTRA_DIRECTIVES_DOCUMENT);
}
exports.getAppSyncServiceExtraDirectives = getAppSyncServiceExtraDirectives;
