// amplify-java-function-runtime-provider/lib/utils/detect.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const which = __importStar(require("which"));
const execa = __importStar(require("execa"));
const semver = __importStar(require("semver"));
const constants_1 = require("./constants");
exports.checkJava = async () => {
    const executablePath = which.sync('java', {
        nothrow: true,
    });
    if (executablePath === null) {
        return {
            hasRequiredDependencies: false,
            errorMessage: `Unable to find Java version ${constants_1.minJavaVersion} on the path. Download link: https://amzn.to/2UUljp9`,
        };
    }
    const result = execa.sync('java', ['-version']);
    if (result.exitCode !== 0) {
        throw new Error(`java failed, exit code was ${result.exitCode}`);
    }
    const regex = /(\d+\.)(\d+\.)(\d)/g;
    const versionString = result.stderr ? result.stderr.split(/\r?\n/)[0] : '';
    const version = versionString.match(regex);
    if (version !== null && semver.satisfies(version[0], constants_1.minJavaVersion)) {
        return {
            hasRequiredDependencies: true,
        };
    }
    return {
        hasRequiredDependencies: false,
        errorMessage: `Update JDK to ${constants_1.minJavaVersion}. Download link: https://amzn.to/2UUljp9`,
    };
};
exports.checkGradle = async () => {
    const executablePath = which.sync('gradle', {
        nothrow: true,
    });
    if (executablePath === null) {
        return {
            hasRequiredDependencies: false,
            errorMessage: `Unable to find Gradle version ${constants_1.minGradleVersion} on the path. Download link: https://bit.ly/3aGYDj6`,
        };
    }
    const result = execa.sync('gradle', ['-v']);
    if (result.exitCode !== 0) {
        throw new Error(`gradle failed, exit code was ${result.exitCode}`);
    }
    const regex = /(\d+\.)(\d+)/g;
    const versionLines = result.stdout ? result.stdout.split(/\r?\n/) : [];
    const versionString = versionLines.length >= 3 ? versionLines[2] : '';
    const version = versionString.match(regex);
    if (version !== null && semver.satisfies(version[0] + '.0', constants_1.minGradleVersion)) {
        return {
            hasRequiredDependencies: true,
        };
    }
    return {
        hasRequiredDependencies: false,
        errorMessage: `Update Gradle to ${constants_1.minGradleVersion}. Download link: https://bit.ly/3aGYDj6`,
    };
};
exports.checkJavaCompiler = async () => {
    const executablePath = which.sync('javac', {
        nothrow: true,
    });
    if (executablePath === null) {
        return {
            hasRequiredDependencies: false,
            errorMessage: `Unable to find Java compiler version ${constants_1.minJavaVersion} on the path. Download link: https://amzn.to/2UUljp9`,
        };
    }
    const result = execa.sync('javac', ['-version']);
    if (result.exitCode !== 0) {
        throw new Error(`java failed, exit code was ${result.exitCode}`);
    }
    const regex = /(\d+\.)(\d+\.)(\d)/g;
    const versionString = result.stdout ? result.stdout.split(/\r?\n/)[0] : '';
    const version = versionString.match(regex);
    if (version !== null && semver.satisfies(version[0], constants_1.minJavaVersion)) {
        return {
            hasRequiredDependencies: true,
        };
    }
    return {
        hasRequiredDependencies: false,
        errorMessage: `Update JDK to ${constants_1.minJavaVersion}. Download link: https://amzn.to/2UUljp9`,
    };
};
