// amplify-nodejs-function-runtime-provider/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.functionRuntimeContributorFactory = void 0;
const legacyBuild_1 = require("./utils/legacyBuild");
const legacyPackage_1 = require("./utils/legacyPackage");
const invoke_1 = require("./utils/invoke");
const path_1 = __importDefault(require("path"));
exports.functionRuntimeContributorFactory = context => {
    return {
        contribute: request => {
            const selection = request.selection;
            if (selection !== 'nodejs') {
                return Promise.reject(new Error(`Unknown selection ${selection}`));
            }
            return Promise.resolve({
                runtime: {
                    name: 'NodeJS',
                    value: 'nodejs',
                    cloudTemplateValue: 'nodejs12.x',
                    defaultHandler: 'index.handler',
                },
            });
        },
        checkDependencies: () => Promise.resolve({ hasRequiredDependencies: true }),
        package: params => legacyPackage_1.packageResource(params, context),
        build: params => legacyBuild_1.buildResource(params),
        invoke: async (params) => {
            await legacyBuild_1.buildResource(params);
            return invoke_1.invoke({
                packageFolder: path_1.default.join(params.srcRoot, 'src'),
                handler: params.handler,
                event: params.event,
                environment: params.envVars,
            });
        },
    };
};
