// amplify-python-function-runtime-provider/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const buildUtils_1 = require("./util/buildUtils");
const packageUtils_1 = require("./util/packageUtils");
const invokeUtil_1 = require("./util/invokeUtil");
const depUtils_1 = require("./util/depUtils");
exports.functionRuntimeContributorFactory = context => {
    return {
        contribute: request => {
            const selection = request.selection;
            if (selection !== 'python') {
                return Promise.reject(new Error(`Unknown selection ${selection}`));
            }
            return Promise.resolve({
                runtime: {
                    name: 'Python',
                    value: 'python',
                    cloudTemplateValue: 'python3.8',
                    defaultHandler: 'index.handler',
                },
            });
        },
        checkDependencies: depUtils_1.checkDeps,
        package: request => packageUtils_1.pythonPackage(context, request),
        build: buildUtils_1.pythonBuild,
        invoke: request => buildUtils_1.pythonBuild(request).then(() => invokeUtil_1.pythonInvoke(context, request)),
    };
};
