// graphql-relational-schema-transformer/lib/ResourceConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceConstants = void 0;
let ResourceConstants = (() => {
    class ResourceConstants {
    }
    ResourceConstants.ENVIRONMENT_CONTEXT_KEYS = {
        RDSRegion: 'rdsRegion',
        RDSClusterIdentifier: 'rdsClusterIdentifier',
        RDSSecretStoreArn: 'rdsSecretStoreArn',
        RDSDatabaseName: 'rdsDatabaseName',
    };
    ResourceConstants.RESOURCES = {
        GraphQLAPILogicalID: 'GraphQLAPI',
        GraphQLSchemaLogicalID: 'GraphQLSchema',
        APIKeyLogicalID: 'GraphQLAPIKey',
        ResolverFileName: 'ResolverFileName',
        RelationalDatabaseDataSource: 'RelationalDatabaseDataSource',
        RelationalDatabaseAccessRole: 'RelationalDatabaseAccessRole',
    };
    ResourceConstants.PARAMETERS = {
        Env: 'env',
        S3DeploymentBucket: 'S3DeploymentBucket',
        S3DeploymentRootKey: 'S3DeploymentRootKey',
        AppSyncApiName: 'AppSyncApiName',
        AppSyncApiId: 'AppSyncApiId',
        APIKeyExpirationEpoch: 'APIKeyExpirationEpoch',
        rdsRegion: 'rdsRegion',
        rdsClusterIdentifier: 'rdsClusterIdentifier',
        rdsSecretStoreArn: 'rdsSecretStoreArn',
        rdsDatabaseName: 'rdsDatabaseName',
    };
    return ResourceConstants;
})();
exports.ResourceConstants = ResourceConstants;
