// amplify-python-function-template-provider/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = __importDefault(require("fs-extra"));
const pathToTemplateFiles = `${__dirname}/../resources/hello-world`;
exports.functionTemplateContributorFactory = context => {
    return {
        contribute: request => {
            const selection = request.selection;
            if (selection !== 'hello-world') {
                throw new Error(`Unknown python template selection ${selection}`);
            }
            return helloWorld();
        },
    };
};
function helloWorld() {
    const files = fs_extra_1.default.readdirSync(pathToTemplateFiles);
    return Promise.resolve({
        functionTemplate: {
            sourceRoot: pathToTemplateFiles,
            sourceFiles: files,
            destMap: {
                'index.py': 'src/index.py',
                'event.json': 'src/event.json',
            },
            defaultEditorFile: 'src/index.py',
        },
    });
}
exports.helloWorld = helloWorld;
