// @aws-amplify/cli/lib/plugin-helpers/access-plugins-file.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.writePluginsJsonFile = exports.writePluginsJsonFileSync = exports.readPluginsJsonFile = exports.readPluginsJsonFileSync = void 0;
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const constants_1 = require("../domain/constants");
const readJsonFile_1 = require("../utils/readJsonFile");
const JSON_SPACE = 4;
function readPluginsJsonFileSync() {
    let result;
    const pluginsFilePath = getPluginsJsonFilePath();
    if (fs_extra_1.default.existsSync(pluginsFilePath)) {
        result = readJsonFile_1.readJsonFileSync(pluginsFilePath);
    }
    return result;
}
exports.readPluginsJsonFileSync = readPluginsJsonFileSync;
async function readPluginsJsonFile() {
    let result;
    const pluginsFilePath = getPluginsJsonFilePath();
    const exists = await fs_extra_1.default.pathExists(pluginsFilePath);
    if (exists) {
        result = await readJsonFile_1.readJsonFile(pluginsFilePath);
    }
    return result;
}
exports.readPluginsJsonFile = readPluginsJsonFile;
function writePluginsJsonFileSync(pluginsJson) {
    const systemDotAmplifyDirPath = getSystemDotAmplifyDirPath();
    const pluginsJsonFilePath = path_1.default.join(systemDotAmplifyDirPath, getPluginsJsonFileName());
    fs_extra_1.default.ensureDirSync(systemDotAmplifyDirPath);
    const jsonString = JSON.stringify(pluginsJson, null, JSON_SPACE);
    fs_extra_1.default.writeFileSync(pluginsJsonFilePath, jsonString, 'utf8');
}
exports.writePluginsJsonFileSync = writePluginsJsonFileSync;
async function writePluginsJsonFile(pluginsJson) {
    const systemDotAmplifyDirPath = getSystemDotAmplifyDirPath();
    const pluginsJsonFilePath = path_1.default.join(systemDotAmplifyDirPath, getPluginsJsonFileName());
    await fs_extra_1.default.ensureDir(systemDotAmplifyDirPath);
    const jsonString = JSON.stringify(pluginsJson, null, JSON_SPACE);
    await fs_extra_1.default.writeFile(pluginsJsonFilePath, jsonString, 'utf8');
}
exports.writePluginsJsonFile = writePluginsJsonFile;
function getPluginsJsonFilePath() {
    return path_1.default.join(getSystemDotAmplifyDirPath(), getPluginsJsonFileName());
}
function getSystemDotAmplifyDirPath() {
    return path_1.default.join(os_1.default.homedir(), constants_1.constants.DotAmplifyDirName);
}
function getPluginsJsonFileName() {
    let result = constants_1.constants.PLUGINS_FILE_NAME;
    const amplifyExecutableName = path_1.default.basename(process.argv[1]);
    if (amplifyExecutableName === 'amplify-dev') {
        result = `${amplifyExecutableName}-${constants_1.constants.PLUGINS_FILE_NAME}`;
    }
    return result;
}
