// amplify-dotnet-function-runtime-provider/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("./constants");
const detect_1 = require("./utils/detect");
const build_1 = require("./utils/build");
const package_1 = require("./utils/package");
const invoke_1 = require("./utils/invoke");
exports.functionRuntimeContributorFactory = (context) => {
    return {
        checkDependencies: detect_1.detectDotNetCore,
        contribute: async (contributionRequest) => {
            switch (contributionRequest.selection) {
                case constants_1.dotnetcore31:
                    return {
                        runtime: {
                            name: '.NET Core 3.1',
                            value: constants_1.dotnetcore31,
                            cloudTemplateValue: constants_1.dotnetcore31,
                            defaultHandler: `${contributionRequest.contributionContext.resourceName}::${contributionRequest.contributionContext.resourceName}.${contributionRequest.contributionContext.functionName}::LambdaHandler`,
                        },
                    };
                default:
                    throw new Error(`Unknown selection ${contributionRequest.selection}`);
            }
        },
        package: async (request) => package_1.packageAssemblies(request, context),
        build: build_1.build,
        invoke: invoke_1.invoke,
    };
};
