// amplify-codegen-appsync-model-plugin/lib/configs/java-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONNECTION_RELATIONSHIP_IMPORTS = exports.LOADER_IMPORT_PACKAGES = exports.NON_MODEL_CLASS_IMPORT_PACKAGES = exports.ENUM_IMPORT_PACKAGES = exports.MODEL_CLASS_IMPORT_PACKAGES = exports.LOADER_CLASS_NAME = exports.GENERATED_PACKAGE_NAME = void 0;
exports.GENERATED_PACKAGE_NAME = 'com.amplifyframework.datastore.generated.model';
exports.LOADER_CLASS_NAME = 'AmplifyModelProvider';
exports.MODEL_CLASS_IMPORT_PACKAGES = [
    'java.util.List',
    'java.util.UUID',
    'java.util.Objects',
    '',
    'androidx.core.util.ObjectsCompat',
    '',
    'com.amplifyframework.core.model.Model',
    'com.amplifyframework.core.model.annotations.Index',
    'com.amplifyframework.core.model.annotations.ModelConfig',
    'com.amplifyframework.core.model.annotations.ModelField',
    'com.amplifyframework.core.model.query.predicate.QueryField',
    '',
    'static com.amplifyframework.core.model.query.predicate.QueryField.field',
    '',
];
exports.ENUM_IMPORT_PACKAGES = ['com.amplifyframework.core.model.ModelEnum;', ''];
exports.NON_MODEL_CLASS_IMPORT_PACKAGES = ['androidx.core.util.ObjectsCompat', '', 'java.util.Objects', 'java.util.List', ''];
exports.LOADER_IMPORT_PACKAGES = [
    'com.amplifyframework.util.Immutable',
    'com.amplifyframework.core.model.Model',
    'com.amplifyframework.core.model.ModelProvider',
    '',
    'java.util.Arrays',
    'java.util.HashSet',
    'java.util.Set',
];
exports.CONNECTION_RELATIONSHIP_IMPORTS = {
    BELONGS_TO: 'com.amplifyframework.core.model.annotations.BelongsTo',
    HAS_MANY: 'com.amplifyframework.core.model.annotations.HasMany',
    HAS_ONE: 'com.amplifyframework.core.model.annotations.HasOne',
};
