// @aws-amplify/cli/lib/commands/plugin/remove.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = void 0;
const plugin_manager_1 = require("../../plugin-manager");
const constants_1 = require("../../domain/constants");
const inquirer_helper_1 = __importStar(require("../../domain/inquirer-helper"));
async function run(context) {
    const options = new Array();
    const { plugins } = context.pluginPlatform;
    if (plugins && Object.keys(plugins).length > 0) {
        Object.keys(plugins).forEach(key => {
            if (key === constants_1.constants.CORE) {
                return;
            }
            if (plugins[key].length > 0) {
                const option = {
                    name: key + inquirer_helper_1.EXPAND,
                    value: plugins[key],
                    short: key + inquirer_helper_1.EXPAND,
                };
                if (plugins[key].length === 1) {
                    const pluginInfo = plugins[key][0];
                    option.name = pluginInfo.packageName + '@' + pluginInfo.packageVersion;
                    option.short = pluginInfo.packageName + '@' + pluginInfo.packageVersion;
                }
                options.push(option);
            }
        });
    }
    if (options.length > 0) {
        const { selections } = await inquirer_helper_1.default.prompt({
            type: 'checkbox',
            name: 'selections',
            message: 'Select the plugin packages to remove',
            choices: options,
        });
        if (selections.length > 0) {
            const sequential = require('promise-sequential');
            const removeTasks = selections.map((selection) => async () => {
                await removeNamedPlugins(context.pluginPlatform, selection);
            });
            await sequential(removeTasks);
            await plugin_manager_1.confirmAndScan(context.pluginPlatform);
        }
    }
    else {
        context.print.console.error('No plugins are found');
    }
}
exports.run = run;
async function removeNamedPlugins(pluginPlatform, pluginInfos) {
    if (pluginInfos.length === 1) {
        plugin_manager_1.removePluginPackage(pluginPlatform, pluginInfos[0]);
    }
    else if (pluginInfos.length > 1) {
        const options = pluginInfos.map((pluginInfo) => {
            const optionObject = {
                name: pluginInfo.packageName + '@' + pluginInfo.packageVersion,
                value: pluginInfo,
                short: pluginInfo.packageName + '@' + pluginInfo.packageVersion,
            };
            return optionObject;
        });
        const { selections } = await inquirer_helper_1.default.prompt({
            type: 'checkbox',
            name: 'selections',
            message: 'Select the plugin packages to remove',
            choices: options,
        });
        if (selections.length > 0) {
            const sequential = require('promise-sequential');
            const removeTasks = selections.map((pluginInfo) => async () => {
                await plugin_manager_1.removePluginPackage(pluginPlatform, pluginInfo);
            });
            await sequential(removeTasks);
        }
    }
}
