// amplify-appsync-simulator/lib/server/subscription/websocket-server/server.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var WebSocket = __importStar(require("ws"));
var ws_1 = require("ws");
var message_type_guards_1 = require("./message-type-guards");
var message_types_1 = require("./message-types");
var utils_1 = require("./utils");
var PROTOCOL = 'graphql-ws';
var KEEP_ALIVE_TIMEOUT = 4 * 60 * 1000; // Wait time between Keep Alive Message
// Max time the client will wait for Keep Alive message before disconnecting. Sent to the client as part of connection ack
var CONNECTION_TIMEOUT_DURATION = 5 * 60 * 1000;
var DEFAULT_OPTIONS = {
    onConnectHandler: function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
        return [2 /*return*/];
    }); }); },
    keepAlive: KEEP_ALIVE_TIMEOUT,
    connectionTimeoutDuration: CONNECTION_TIMEOUT_DURATION,
};
var WebsocketSubscriptionServer = /** @class */ (function () {
    function WebsocketSubscriptionServer(options, server) {
        var _this = this;
        this.onClose = function (connectionContext) {
            connectionContext.subscriptions.forEach(function (subscriptionId) {
                _this.stopAsyncIterator(connectionContext, subscriptionId.id);
            });
            if (connectionContext.pingIntervalHandle) {
                clearInterval(connectionContext.pingIntervalHandle);
                connectionContext.pingIntervalHandle = null;
            }
            _this.connections.delete(connectionContext);
        };
        this.onUnsubscribe = function (connectionContext, messageOrSubscriptionId) {
            var id = messageOrSubscriptionId.id;
            _this.stopAsyncIterator(connectionContext, id);
            _this.sendMessage(connectionContext, id, message_types_1.MESSAGE_TYPES.GQL_COMPLETE, {});
        };
        this.stopAsyncIterator = function (connectionContext, id) {
            if (connectionContext.subscriptions && connectionContext.subscriptions.has(id)) {
                var subscription = connectionContext.subscriptions.get(id);
                if (subscription.asyncIterator) {
                    subscription.asyncIterator.return();
                }
                connectionContext.subscriptions.delete(id);
            }
        };
        this.onSocketConnection = function (socket, request) { return __awaiter(_this, void 0, void 0, function () {
            var connectionContext_1, headers, onMessage_1, onClose_1, e_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        socket.upgradeReq = request;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        if (typeof socket.protocol === 'undefined' || socket.protocol !== PROTOCOL) {
                            throw new Error('Invalid protocol');
                        }
                        connectionContext_1 = {
                            request: request,
                            socket: socket,
                            subscriptions: new Map(),
                            isConnectionInitialized: false,
                        };
                        headers = utils_1.decodeHeaderFromQueryParam(request.url);
                        return [4 /*yield*/, this.options.onConnectHandler(connectionContext_1, headers)];
                    case 2:
                        _a.sent();
                        this.connections.add(connectionContext_1);
                        onMessage_1 = function (message) {
                            _this.onMessage(connectionContext_1, message);
                        };
                        onClose_1 = function (error) {
                            socket.off('message', onMessage_1);
                            socket.off('close', onClose_1);
                            socket.off('error', onClose_1);
                            _this.onSocketDisconnection(connectionContext_1, error);
                        };
                        socket.on('message', onMessage_1);
                        socket.on('close', onClose_1);
                        socket.on('error', onClose_1);
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        console.log(e_1);
                        socket.close(1002); // protocol error
                        return [2 /*return*/];
                    case 4: return [2 /*return*/];
                }
            });
        }); };
        this.onSocketDisconnection = function (connectionContext, error) {
            _this.onClose(connectionContext);
            if (error) {
                _this.sendError(connectionContext, '', { message: error instanceof Error ? error.message : error });
                setTimeout(function () {
                    // 1011 is an unexpected condition prevented the request from being fulfilled
                    connectionContext.socket.close(1011);
                }, 10);
            }
        };
        this.onMessage = function (connectionContext, rawMessage) {
            var message = JSON.parse(rawMessage);
            try {
                switch (message.type) {
                    case message_types_1.MESSAGE_TYPES.GQL_CONNECTION_INIT:
                        if (message_type_guards_1.isSubscriptionConnectionInitMessage(message)) {
                            return _this.onConnectionInit(connectionContext, message);
                        }
                        break;
                    case message_types_1.MESSAGE_TYPES.GQL_START:
                        if (message_type_guards_1.isSubscriptionStartMessage(message)) {
                            return _this.onSubscriptionStart(connectionContext, message);
                        }
                        break;
                    case message_types_1.MESSAGE_TYPES.GQL_STOP:
                        if (message_type_guards_1.isSubscriptionStopMessage(message)) {
                            return _this.onUnsubscribe(connectionContext, message);
                        }
                }
                throw new Error('Invalid message');
            }
            catch (e) {
                _this.sendError(connectionContext, '', { errors: [{ message: e.message }] });
            }
        };
        this.sendMessage = function (connectionContext, subscriptionId, type, data) {
            var message = {
                type: type,
                id: subscriptionId,
                payload: data,
            };
            if (connectionContext.socket.readyState === WebSocket.OPEN) {
                connectionContext.socket.send(JSON.stringify(message));
            }
        };
        this.sendError = function (connectionContext, subscriptionId, errorPayload, type) {
            if (type === void 0) { type = message_types_1.MESSAGE_TYPES.GQL_ERROR; }
            if ([message_types_1.MESSAGE_TYPES.GQL_CONNECTION_ERROR, message_types_1.MESSAGE_TYPES.GQL_ERROR].indexOf(type) === -1) {
                throw new Error("Message type should for error should be one of " + message_types_1.MESSAGE_TYPES.GQL_ERROR + " or " + message_types_1.MESSAGE_TYPES.GQL_CONNECTION_ERROR);
            }
            _this.sendMessage(connectionContext, subscriptionId, type, errorPayload);
        };
        this.setupPing = function (connectionContext) {
            connectionContext.pingIntervalHandle = setInterval(function () {
                _this.sendMessage(connectionContext, undefined, message_types_1.MESSAGE_TYPES.GQL_CONNECTION_KEEP_ALIVE, undefined);
            }, _this.options.keepAlive);
        };
        this.onConnectionInit = function (connectionContext, message) {
            connectionContext.isConnectionInitialized = true;
            var response = {
                type: message_types_1.MESSAGE_TYPES.GQL_CONNECTION_ACK,
                payload: {
                    connectionTimeout: _this.options.connectionTimeoutDuration,
                },
            };
            _this.sendMessage(connectionContext, undefined, response.type, response.payload);
            _this.setupPing(connectionContext);
            // Regular keep alive messages if keepAlive is set
            _this.setupPing(connectionContext);
        };
        this.onSubscriptionStart = function (connectionContext, message) { return __awaiter(_this, void 0, void 0, function () {
            var id, data, query, variables, headers, asyncIterator, error, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        id = message.id;
                        data = JSON.parse(message.payload.data);
                        query = graphql_1.parse(data.query);
                        variables = data.variables;
                        headers = message.payload.extensions.authorization;
                        if (connectionContext.subscriptions && connectionContext.subscriptions.has(id)) {
                            this.stopAsyncIterator(connectionContext, id);
                        }
                        return [4 /*yield*/, this.options.onSubscribeHandler(query, variables, headers, connectionContext.request)];
                    case 1:
                        asyncIterator = _a.sent();
                        if (asyncIterator.errors) {
                            error = {
                                errors: asyncIterator.errors,
                                data: asyncIterator.data || null,
                            };
                            this.sendError(connectionContext, id, error, message_types_1.MESSAGE_TYPES.GQL_ERROR);
                        }
                        else {
                            subscription = {
                                id: id,
                                asyncIterator: asyncIterator,
                                document: query,
                                variables: variables,
                            };
                            connectionContext.subscriptions.set(id, subscription);
                            this.sendMessage(connectionContext, id, message_types_1.MESSAGE_TYPES.GQL_START_ACK, {});
                            this.attachAsyncIterator(connectionContext, subscription);
                        }
                        return [2 /*return*/];
                }
            });
        }); };
        this.attachAsyncIterator = function (connectionContext, sub) { return __awaiter(_this, void 0, void 0, function () {
            var asyncIterator, id, _a, value, done;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        asyncIterator = sub.asyncIterator, id = sub.id;
                        _b.label = 1;
                    case 1:
                        if (!true) return [3 /*break*/, 3];
                        return [4 /*yield*/, asyncIterator.next()];
                    case 2:
                        _a = _b.sent(), value = _a.value, done = _a.done;
                        if (done) {
                            return [3 /*break*/, 3];
                        }
                        this.sendMessage(connectionContext, id, message_types_1.MESSAGE_TYPES.GQL_DATA, value);
                        return [3 /*break*/, 1];
                    case 3: return [2 /*return*/];
                }
            });
        }); };
        this.connections = new Set();
        this.options = __assign(__assign({}, DEFAULT_OPTIONS), options);
        if (server) {
            this.attachWebServer(server);
        }
    }
    WebsocketSubscriptionServer.prototype.attachWebServer = function (serverOptions) {
        this.webSocketServer = new ws_1.Server(serverOptions || {});
    };
    WebsocketSubscriptionServer.prototype.start = function () {
        if (!this.webSocketServer) {
            throw new Error('No server is attached');
        }
        this.webSocketServer.on('connection', this.onSocketConnection);
    };
    WebsocketSubscriptionServer.prototype.stop = function () {
        var _this = this;
        if (this.webSocketServer) {
            this.webSocketServer.off('connection', this.onSocketConnection);
            this.connections.forEach(function (connection) {
                _this.onClose(connection);
            });
            this.webSocketServer.close();
        }
    };
    return WebsocketSubscriptionServer;
}());
exports.WebsocketSubscriptionServer = WebsocketSubscriptionServer;
