// amplify-category-function/lib/provider-utils/awscloudformation/service-walkthroughs/execPermissionsWalkthrough.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.askExecRolePermissionsQuestions = void 0;
const cloudformationHelpers_1 = require("../utils/cloudformationHelpers");
const inquirer_1 = __importDefault(require("inquirer"));
const path_1 = __importDefault(require("path"));
const TransformPackage = __importStar(require("graphql-transformer-core"));
const constants_1 = require("../../../constants");
async function askExecRolePermissionsQuestions(context, allDefaultValues, parameters, currentDefaults) {
    const amplifyMetaFilePath = context.amplify.pathManager.getAmplifyMetaFilePath();
    const amplifyMeta = context.amplify.readJsonFile(amplifyMetaFilePath);
    let categories = Object.keys(amplifyMeta);
    categories = categories.filter(category => category !== 'providers');
    const appsyncResourceName = 'api' in amplifyMeta ? Object.keys(amplifyMeta.api).find(key => amplifyMeta.api[key].service === 'AppSync') : undefined;
    if (!categories.includes('storage') && appsyncResourceName !== undefined) {
        categories.push('storage');
    }
    const categoryPermissionQuestion = {
        type: 'checkbox',
        name: 'categories',
        message: 'Select the category',
        choices: categories,
        default: currentDefaults ? currentDefaults.categories : undefined,
    };
    const capitalizeFirstLetter = str => str.charAt(0).toUpperCase() + str.slice(1);
    const categoryPermissionAnswer = await inquirer_1.default.prompt([categoryPermissionQuestion]);
    const selectedCategories = categoryPermissionAnswer.categories;
    let categoryPolicies = [];
    let resources = [];
    const crudOptions = ['create', 'read', 'update', 'delete'];
    parameters.permissions = {};
    const backendDir = context.amplify.pathManager.getBackendDirPath();
    const appsyncTableSuffix = '@model(appsync)';
    for (let category of selectedCategories) {
        const resourcesList = category in amplifyMeta ? Object.keys(amplifyMeta[category]) : [];
        if (category === 'storage' && 'api' in amplifyMeta) {
            if (appsyncResourceName) {
                const resourceDirPath = path_1.default.join(backendDir, 'api', appsyncResourceName);
                const project = await TransformPackage.readProjectConfiguration(resourceDirPath);
                const directivesMap = TransformPackage.collectDirectivesByTypeNames(project.schema);
                const modelNames = Object.keys(directivesMap.types)
                    .filter(typeName => directivesMap.types[typeName].includes('model'))
                    .map(modelName => `${modelName}:${appsyncTableSuffix}`);
                resourcesList.push(...modelNames);
            }
        }
        if (resourcesList.length === 0) {
            context.print.warning(`No resources found for ${category}`);
            continue;
        }
        try {
            let selectedResources = [];
            if (resourcesList.length === 1) {
                context.print.info(`${capitalizeFirstLetter(category)} category has a resource called ${resourcesList[0]}`);
                selectedResources = [resourcesList[0]];
            }
            else {
                const resourceQuestion = {
                    type: 'checkbox',
                    name: 'resources',
                    message: `${capitalizeFirstLetter(category)} has ${resourcesList.length} resources in this project. Select the one you would like your Lambda to access`,
                    choices: resourcesList,
                    validate: value => {
                        if (value.length === 0) {
                            return 'You must select at least resource';
                        }
                        return true;
                    },
                    default: () => {
                        if (currentDefaults && currentDefaults.categoryPermissionMap && currentDefaults.categoryPermissionMap[category]) {
                            return Object.keys(currentDefaults.categoryPermissionMap[category]);
                        }
                    },
                };
                const resourceAnswer = await inquirer_1.default.prompt([resourceQuestion]);
                selectedResources = resourceAnswer.resources;
            }
            for (let resourceName of selectedResources) {
                const pluginInfo = context.amplify.getCategoryPluginInfo(context, category, resourceName);
                const { getPermissionPolicies } = await Promise.resolve().then(() => __importStar(require(pluginInfo.packageLocation)));
                if (!getPermissionPolicies) {
                    context.print.warning(`Policies cannot be added for ${category}/${resourceName}`);
                    continue;
                }
                else {
                    const crudPermissionQuestion = {
                        type: 'checkbox',
                        name: 'crudOptions',
                        message: `Select the operations you want to permit for ${resourceName}`,
                        choices: crudOptions,
                        validate: value => {
                            if (value.length === 0) {
                                return 'You must select at least one operation';
                            }
                            return true;
                        },
                        default: () => {
                            if (currentDefaults &&
                                currentDefaults.categoryPermissionMap &&
                                currentDefaults.categoryPermissionMap[category] &&
                                currentDefaults.categoryPermissionMap[category][resourceName]) {
                                return currentDefaults.categoryPermissionMap[category][resourceName];
                            }
                        },
                    };
                    const crudPermissionAnswer = await inquirer_1.default.prompt([crudPermissionQuestion]);
                    const resourcePolicy = crudPermissionAnswer.crudOptions;
                    if (resourceName.endsWith(appsyncTableSuffix)) {
                        resourcePolicy.providerPlugin = 'awscloudformation';
                        resourcePolicy.service = 'DynamoDB';
                        const dynamoDBTableARNComponents = cloudformationHelpers_1.constructCFModelTableArnComponent(appsyncResourceName, resourceName, appsyncTableSuffix);
                        resourcePolicy.customPolicyResource = [
                            {
                                'Fn::Join': ['', dynamoDBTableARNComponents],
                            },
                            {
                                'Fn::Join': ['', [...dynamoDBTableARNComponents, '/index/*']],
                            },
                        ];
                    }
                    const { permissionPolicies, resourceAttributes } = await getPermissionPolicies(context, { [resourceName]: resourcePolicy });
                    categoryPolicies = categoryPolicies.concat(permissionPolicies);
                    if (!parameters.permissions[category]) {
                        parameters.permissions[category] = {};
                    }
                    parameters.permissions[category][resourceName] = resourcePolicy;
                    resources = resources.concat(resourceAttributes.map(attributes => attributes.resourceName && attributes.resourceName.endsWith(appsyncTableSuffix)
                        ? {
                            resourceName: appsyncResourceName,
                            category: 'api',
                            attributes: ['GraphQLAPIIdOutput'],
                            needsAdditionalDynamoDBResourceProps: true,
                            _modelName: attributes.resourceName.replace(`:${appsyncTableSuffix}`, 'Table'),
                            _cfJoinComponentTableName: cloudformationHelpers_1.constructCFModelTableNameComponent(appsyncResourceName, attributes.resourceName, appsyncTableSuffix),
                            _cfJoinComponentTableArn: cloudformationHelpers_1.constructCFModelTableArnComponent(appsyncResourceName, attributes.resourceName, appsyncTableSuffix),
                        }
                        : attributes));
                }
            }
        }
        catch (e) {
            context.print.warning(`Policies cannot be added for ${category}`);
            context.print.info(e.stack);
        }
    }
    allDefaultValues.categoryPolicies = categoryPolicies;
    const resourceProperties = [];
    const resourcePropertiesJSON = {};
    const envVars = new Set();
    allDefaultValues.dependsOn = [];
    resources.forEach(resource => {
        const { category, resourceName, attributes } = resource;
        if (resource.needsAdditionalDynamoDBResourceProps) {
            const modelEnvPrefix = `${category.toUpperCase()}_${resourceName.toUpperCase()}_${resource._modelName.toUpperCase()}`;
            let modelArnResourcePropValue = {
                'Fn::Join': ['', resource._cfJoinComponentTableArn],
            };
            resourceProperties.push(`"${modelEnvPrefix}_NAME": ${JSON.stringify(resource._cfJoinComponentTableName)}`);
            resourceProperties.push(`"${modelEnvPrefix}_ARN": ${JSON.stringify(modelArnResourcePropValue)}`);
            resourcePropertiesJSON[`${modelEnvPrefix}_NAME`] = resource._cfJoinComponentTableName;
            resourcePropertiesJSON[`${modelEnvPrefix}_ARN`] = modelArnResourcePropValue;
            envVars.add(`${modelEnvPrefix}_NAME`);
            envVars.add(`${modelEnvPrefix}_ARN`);
        }
        attributes.forEach(attribute => {
            const envName = `${category.toUpperCase()}_${resourceName.toUpperCase()}_${attribute.toUpperCase()}`;
            const refName = `${category}${resourceName}${attribute}`;
            resourceProperties.push(`"${envName}": {"Ref": "${refName}"}`);
            resourcePropertiesJSON[`${envName}`] = { Ref: refName };
            envVars.add(envName);
        });
        let resourceExists = false;
        allDefaultValues.dependsOn.forEach(amplifyResource => {
            if (amplifyResource.resourceName === resourceName) {
                resourceExists = true;
            }
        });
        if (!resourceExists) {
            allDefaultValues.dependsOn.push({
                category: resource.category,
                resourceName: resource.resourceName,
                attributes: resource.attributes,
            });
        }
    });
    allDefaultValues.resourceProperties = resourceProperties.join(',\n');
    allDefaultValues.resourcePropertiesJSON = resourcePropertiesJSON;
    envVars.add('ENV');
    envVars.add('REGION');
    const envVarStringList = Array.from(envVars)
        .sort()
        .join('\n\t');
    context.print.info(`${constants_1.envVarPrintoutPrefix}${envVarStringList}`);
    return { topLevelComment: `${constants_1.topLevelCommentPrefix}${envVarStringList}${constants_1.topLevelCommentSuffix}` };
}
exports.askExecRolePermissionsQuestions = askExecRolePermissionsQuestions;
