// amplify-category-function/lib/provider-utils/awscloudformation/service-walkthroughs/generalQuestionsWalkthrough.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const inquirer_1 = __importDefault(require("inquirer"));
const uuid_1 = __importDefault(require("uuid"));
async function generalQuestionsWalkthrough(context) {
    return await inquirer_1.default.prompt(generalQuestions(context));
}
exports.default = generalQuestionsWalkthrough;
function generalQuestions(context) {
    return [
        {
            type: 'input',
            name: 'resourceName',
            message: 'Provide a friendly name for your resource to be used as a label for this category in the project:',
            validate: context.amplify.inputValidation({
                operator: 'regex',
                value: '^[a-zA-Z0-9]+$',
                onErrorMsg: 'Resource name should be alphanumeric',
                required: true,
            }),
            default: () => {
                const appName = context.amplify
                    .getProjectDetails()
                    .projectConfig.projectName.toLowerCase()
                    .replace(/[^0-9a-zA-Z]/gi, '');
                const [shortId] = uuid_1.default().split('-');
                return `${appName}${shortId}`;
            },
        },
        {
            type: 'input',
            name: 'functionName',
            message: 'Provide the AWS Lambda function name:',
            validate: context.amplify.inputValidation({
                operator: 'regex',
                value: '^[a-zA-Z0-9._-]+$',
                onErrorMsg: 'You can use the following characters: a-z A-Z 0-9 . - _',
                required: true,
            }),
            default: answers => answers.resourceName,
        },
    ];
}
