// amplify-go-function-runtime-provider/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const runtime_1 = require("./runtime");
const localinvoke_1 = require("./localinvoke");
exports.functionRuntimeContributorFactory = context => {
    return {
        contribute: request => {
            if (request.selection !== 'go') {
                return Promise.reject(new Error(`Unknown selection ${request.selection}`));
            }
            return Promise.resolve({
                runtime: {
                    name: 'Go 1.x',
                    value: 'go1.x',
                    cloudTemplateValue: 'go1.x',
                    defaultHandler: 'main',
                },
            });
        },
        checkDependencies: runtimeValue => runtime_1.checkDependencies(runtimeValue),
        package: request => runtime_1.packageResource(request, context),
        build: request => runtime_1.buildResource(request, context),
        invoke: request => localinvoke_1.localInvoke(request, context),
    };
};
