// amplify-java-function-runtime-provider/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const build_1 = require("./utils/build");
const package_1 = require("./utils/package");
const detect_1 = require("./utils/detect");
const invoke_1 = require("./utils/invoke");
exports.functionRuntimeContributorFactory = context => {
    return {
        contribute: request => {
            const selection = request.selection;
            if (selection !== 'java') {
                return Promise.reject(new Error(`Unknown selection ${selection}`));
            }
            return Promise.resolve({
                runtime: {
                    name: 'Java',
                    value: 'java',
                    cloudTemplateValue: 'java11',
                    defaultHandler: 'example.LambdaRequestHandler::handleRequest',
                },
            });
        },
        checkDependencies: async () => {
            var result = {
                hasRequiredDependencies: true,
            };
            const resultJava = await detect_1.checkJava();
            const resultCompileJava = await detect_1.checkJavaCompiler();
            const resultGradle = await detect_1.checkGradle();
            const errArray = [];
            if (resultJava.errorMessage !== undefined) {
                errArray.push(resultJava.errorMessage);
            }
            if (resultCompileJava.errorMessage !== undefined) {
                errArray.push(resultCompileJava.errorMessage);
            }
            if (resultGradle.errorMessage !== undefined) {
                errArray.push(resultGradle.errorMessage);
            }
            result.hasRequiredDependencies =
                resultJava.hasRequiredDependencies && resultCompileJava.hasRequiredDependencies && resultGradle.hasRequiredDependencies;
            if (result.hasRequiredDependencies === false) {
                result.errorMessage = errArray.join('\n');
            }
            return result;
        },
        package: params => package_1.packageResource(params, context),
        build: params => build_1.buildResource(params),
        invoke: params => invoke_1.invokeResource(params, context),
    };
};
