// amplify-category-function/lib/provider-utils/awscloudformation/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CronExpressionsMode = exports.provider = exports.parametersFileName = exports.functionParametersFileName = exports.serviceName = exports.categoryName = void 0;
const constants_1 = require("../../../constants");
exports.categoryName = constants_1.category;
exports.serviceName = 'Lambda';
exports.functionParametersFileName = 'function-parameters.json';
exports.parametersFileName = 'parameters.json';
exports.provider = 'awscloudformation';
var CronExpressionsMode;
(function (CronExpressionsMode) {
    CronExpressionsMode["Minutes"] = "Minutes";
    CronExpressionsMode["Hourly"] = "Hourly";
    CronExpressionsMode["Daily"] = "Daily";
    CronExpressionsMode["Weekly"] = "Weekly";
    CronExpressionsMode["Monthly"] = "Monthly";
    CronExpressionsMode["Yearly"] = "Yearly";
    CronExpressionsMode["Custom"] = "Custom AWS cron expression";
})(CronExpressionsMode = exports.CronExpressionsMode || (exports.CronExpressionsMode = {}));
